### ISSUE TITLE
0 Session management - main task [master]

### ISSUE THREAD

**Issue by marlass** [SAP/cloud-commerce-spartacus-storefront#1018]
**TL;DR** Please read it.

### What we want to achieve

We would like to provide the best user experience for every customer.
Big part of this is making "sign in" flow simple.
To communicate with API, first we need to authorize using OAuth protocol.
We have at our disposal 2 flows. Would be good to offer clients option to choose which flow they want to use.

#### First option: Resource Owner Password Flow

      +-------------+
      |             |
      |     User    |
      |             |
      +--+----------+
         |
         |
         | 1. Enters credentials (username/credentials)
         |
         v
      +-------------+   2.Send credentials          +---------------+
      |             +------------------------------>|               |
      | Application |                               |  Auth server  |
      |             |<------------------------------+               |
      +----+--------+   3. Access and return token  +---------------+
           |
           |
           | 4. Use access token for verification
           |
           v
      +-------------+
      |             |
      |     API     |
      |             |
      +-------------+

We currently use that one.
User credentials (username/password) are send along with client_id and secret. In response we get access_token and refresh_token.
For API requests we use access_token. After access_token expires (we would get error status code for API requests), we can use refresh_token to get new access_token.
If refresh_token also expires we need to login again.

**Pros:**

- we are able to have "soft" session
- we have control over login screen

**Cons:**

- website is responsible for keeping user credentials secure (potential library users (or we) might introduce some vulnerability in JS that would allow "bad" JS gather user credentials)
- refresh token must have very limited scope so we able to live with the risk of getting them stolen

**Flow in our code:**

1. Credentials unavailable -> redirect to login page
2. User inputs credentials
3. Pass these credentials for access_token and refresh_token
4. Use access_token for API requests
5. Error saying token is expired
6. Use refresh token to get new access_token (returned access token have limited scope: basic user information/cart)
7. Use new access_token for API requests
8. User want's to invoke "high security" feature (checkout / profile management)
9. API returns information about missing permissions
10. Redirect user to login page to get more privileged access_token (cycle repeats)
11. (After some long time) Refresh token is expired (redirect to login page)

#### Second option: Implicit flow

     +----------+
     | Resource |
     |  Owner   |
     |          |
     +----------+
          ^
          |
         (B)
     +----|-----+          Client Identifier     +---------------+
     |         -+----(A)-- & Redirection URI --->|               |
     |  User-   |                                | Authorization |
     |  Agent  -|----(B)-- User authenticates -->|     Server    |
     |          |                                |               |
     |          |<---(C)--- Redirection URI ----<|               |
     |          |          with Access Token     +---------------+
     |          |            in Fragment
     |          |                                +---------------+
     |          |----(D)--- Redirection URI ---->|   Web-Hosted  |
     |          |          without Fragment      |     Client    |
     |          |                                |    Resource   |
     |     (F)  |<---(E)------- Script ---------<|               |
     |          |                                +---------------+
     +-|--------+
       |    |
      (A)  (G) Access Token
       |    |
       ^    v
     +---------+
     |         |
     |  Client |
     |         |
     +---------+

    (A)  The client initiates the flow by directing the resource owner's
        user-agent to the authorization endpoint.  The client includes
        its client identifier, requested scope, local state, and a
        redirection URI to which the authorization server will send the
        user-agent back once access is granted (or denied).

    (B)  The authorization server authenticates the resource owner (via
        the user-agent) and establishes whether the resource owner
        grants or denies the client's access request.

    (C)  Assuming the resource owner grants access, the authorization
        server redirects the user-agent back to the client using the
        redirection URI provided earlier.  The redirection URI includes
        the access token in the URI fragment.

    (D)  The user-agent follows the redirection instructions by making a
        request to the web-hosted client resource (which does not
        include the fragment per [RFC2616]).  The user-agent retains the
        fragment information locally.

    (E)  The web-hosted client resource returns a web page (typically an
        HTML document with an embedded script) capable of accessing the
        full redirection URI including the fragment retained by the
        user-agent, and extracting the access token (and other
        parameters) contained in the fragment.

    (F)  The user-agent executes the script provided by the web-hosted
        client resource locally, which extracts the access token.

    (G)  The user-agent passes the access token to the client.

**Pros:**

- user credentials are secure
- we don't need to secure refresh_token (because there isn't one)
- recommended way to implement OAuth in SPA

**Cons:**

- we have redirect to default backend login page
- "soft" session can only be done with saving cartId, user details (hacky solution)

**Flow in code:**

1. Credentials unavailable -> redirect to login page
2. User inputs credentials, we get access_token
3. Use access_token for API requests
4. Error saying token is expired
5. Redirect user to login page

### Storing credentials

For now we only want to do SSR without context (better caching). That's why we decided to use localStorage/sessionStorage instead of cookies.

- session storage used for storing access_token (closing tab will start "soft" session)
- local storage used for refresh_token

### Configuration

- most of configuration is done in backoffice, where you set expiration times, scopes
- you can only change flow in lib config (API needs to be thought yet)

### Further ideas

- to better protect user credentials some parts of app can be extracted into sandboxed iframes (login/register/payment details)
- users of spartacus must be aware that credentials can be stolen with some "bad" script included in page (iframes should help with those issues)

### Us and Kyma

We'll need to run this approach by Kyma / Honeybadgers. Bill to organize meeting.

**Comment by marlass** [comment]
After experimenting with OAuth configuration in backoffice I found few things impossible to do in the current version (1811).

Findings of resource owner password flow:
- we don't have enough configuration options regarding scopes. Only ROLE_CLIENT and ROLE_TRUSTED_CLIENT are available. There isn't a role that would restrict API access to only cart/basic user information. Because of that, we can't have soft-login with refresh token, but we would need to work on a hacky solution (storing user details and cartId) to simulate that behavior. This solution would be used both in this flow, as well as in implicit flow.

Findings of implicit flow:
- we can't place auth screen in iframe inside spartacus, because of restriction to sameorigin iframe

**Comment by marlass** [comment]
## Things that have to be implemented in platform for soft login feature

- introduce role or scope that limits user possible actions to cart manipulation/basic information access (ROLE_CLIENT is used for a normal session). A new one should be used for soft session. It can be done as a scope. I would also know the differences between scope and roles.
- allow in one OAuth client to define different role/scope for different authorization method. Example: for authorization with credentials we would like to set scope or role with full access, but for authorization with refresh_token in the same client scope or role would be limited to soft session requirements.

### Issue with Kyma

- for Kyma we require OpenId, which is only available in Implicit Flow. However, we need to use Resource Owner Password Flow to match our requirements. Having the ability to request OpenID in this flow would greatly help. Don't really know security implication about that one, but that would be the easiest option for us.
